### PR TITLE
GH#25132: test: harden prefetch gh read detector

### LINKED ISSUE #25132: quality-debt: .agents/scripts/tests/test-pulse-batch-prefetch-conditional-rest.sh — PR #25124 review feedback (medium)
## Unactioned Review Feedback

**Source PR**: #25124
**File**: `.agents/scripts/tests/test-pulse-batch-prefetch-conditional-rest.sh`
**Reviewers**: gemini
**Findings**: 2
**Max severity**: medium

---

### MEDIUM: gemini (gemini-code-assist[bot])
**File**: `.agents/scripts/tests/test-pulse-batch-prefetch-conditional-rest.sh:220`
medium

The current assertion for detecting raw `gh api` or `gh search` calls has several limitations and edge cases:

1. **Commented-out code**: Any commented-out lines (e.g., `# gh api ...`) will trigger a false positive and fail the test. Pre-filtering comments with `sed` avoids this.
2. **Command Substitution**: It misses common patterns like `export res=$(gh api)`, `readonly res=$(gh api)`, `for repo in $(gh api)`, or spaces inside the parenthesis like `$( gh api)`. Since `$(gh api)` executes the command at runtime, any occurrence of it (even inside double-quoted log strings) is a raw call that should be caught. We can match `\$\([[:space:]]*gh` anywhere on the line to safely and robustly catch all command substitutions without false positives on simple `echo "gh api"` strings.
3. **Direct Execution**: It misses direct execution with a negation operator (e.g., `if ! gh api ...` or `! gh api ...`), within subshells/blocks (e.g., `{ gh api; }` or `(gh api)`), or after pipelines/semicolons (e.g., `some_cmd | gh api`).

We can refine the pipeline and regex to address all of these cases robustly.

```suggestion
	if { sed -E 's/^[[:space:]]*#.*//' "$HELPER" \
		| awk '{ if (sub(/\$/, "")) { printf "%s", $0 } else { print } }' \
		| sed -E 's/_prefetch_gh_read[[:space:]]+gh[[:space:]]+(api|search)//g' \
		| grep -nE '^[[:space:]]*([{(][[:space:]]*)?((if|while|until|then|do|else|elif|!)[[:space:]]+)*gh[[:space:]]+(api|search)([[:space:]]|$)|[;|][[:space:]]*gh[[:space:]]+(api|search)([[:space:]]|$)|\$\([[:space:]]*gh[[:space:]]+(api|search)([[:space:]]|$|\))'; } >/dev/null 2>&1; then
```

[View comment](https://github.com/marcusquinn/aidevops/pull/25124#discussion_r3443840246)

---

### MEDIUM: gemini (gemini-code-assist[bot])
**File**: `.agents/scripts/tests/test-pulse-batch-prefetch-conditional-rest.sh:220`
**Verification**: kept as unverifiable (no stable snippet extracted)
Thank you for the update, @vladimirdulov. The implementation looks robust and addresses the edge cases discussed, including comment filtering and broader detection of `gh` command patterns. Since you have verified the changes with both the test script and `shellcheck`, this looks good to go.

[View comment](https://github.com/marcusquinn/aidevops/pull/25124#discussion_r3443861499)

---

---

_Auto-generated by `quality-feedback-helper.sh scan-merged`. Review each finding and either fix the code or dismiss with a reason._

---
**To approve or decline**, use one of:
- `sudo aidevops approve issue <number> marcusquinn/aidevops` -- cryptographically signs approval for automated dispatch
- Comment `declined: <reason>` -- closes this issue (include your reason after the colon)

---
[aidevops.sh](https://aidevops.sh) v3.20.96 automated scan.

### PR DESCRIPTION
**Pull request by vladimirdulov**
## Summary

Refactored the pulse batch prefetch raw gh read assertion into a reusable detector, added comment filtering and broader shell execution matching, and added regression coverage for negated, grouped, and command-substitution gh api/search calls.

## Files Changed

.agents/scripts/tests/test-pulse-batch-prefetch-conditional-rest.sh

## Runtime Testing

- **Risk level:** Low (agent prompts / infrastructure scripts)
- **Verification:** bash .agents/scripts/tests/test-pulse-batch-prefetch-conditional-rest.sh; shellcheck .agents/scripts/tests/test-pulse-batch-prefetch-conditional-rest.sh

Resolves #25132

---
[aidevops.sh](https://aidevops.sh) v3.20.96 plugin for [OpenCode](https://opencode.ai) v1.17.8 with gpt-5.5 spent 2m and 40,674 tokens on this as a headless worker.

### CODE REVIEW

**Comment by vladimirdulov** [comment]
## Completion Summary

- **What**: Refactored the pulse batch prefetch raw gh read assertion into a reusable detector, added comment filtering and broader shell execution matching, and added regression coverage for negated, grouped, and command-substitution gh api/search calls.
- **Issue**: #25132
- **Files changed**: .agents/scripts/tests/test-pulse-batch-prefetch-conditional-rest.sh
- **Testing**: bash .agents/scripts/tests/test-pulse-batch-prefetch-conditional-rest.sh; shellcheck .agents/scripts/tests/test-pulse-batch-prefetch-conditional-rest.sh
- **Key decisions**: none

---
[aidevops.sh](https://aidevops.sh) v3.20.96 plugin for [OpenCode](https://opencode.ai) v1.17.8 with gpt-5.5 spent 2m and 40,674 tokens on this as a headless worker.

**Review by alex-solovyev — DISMISSED** [review]
Auto-approved by pulse runner @alex-solovyev — author @vladimirdulov confirmed collaborator, pre-merge gates passed.

---
[aidevops.sh](https://aidevops.sh) v3.20.96 automated scan.

**Review by marcusquinn — DISMISSED** [review]
Auto-approved by pulse runner @marcusquinn — author @vladimirdulov confirmed collaborator, pre-merge gates passed.

---
[aidevops.sh](https://aidevops.sh) v3.20.96 automated scan.

File: .agents/scripts/tests/test-pulse-batch-prefetch-conditional-rest.sh [hunk 1]
```diff
@@ -115,6 +115,17 @@ JSON
 	return 0
 }
 
+prefetch_raw_gh_read_matches() {
+	local file_path="$1"
+	if { sed -E 's/^[[:space:]]*#.*//' "$file_path" \
```
[the remark this answers is not in the document]
Comment (vladimirdulov) [reply .agents/scripts/tests/test-pulse-batch-prefetch-conditional-rest.sh:123]: @gemini-code-assist Addressed in 5db42f918.

Evidence:
- `.agents/scripts/tests/test-pulse-batch-prefetch-conditional-rest.sh:119` now uses `${1:-}`.
- `.agents/scripts/tests/test-pulse-batch-prefetch-conditional-rest.sh:120` returns before `sed` when the path is empty or not a file.
- Added focused coverage in `test_prefetch_raw_gh_read_detector_rejects_missing_file`.
- Verified: `bash .agents/scripts/tests/test-pulse-batch-prefetch-conditional-rest.sh` (9/9 pass) and `shellcheck .agents/scripts/tests/test-pulse-batch-prefetch-conditional-rest.sh`.

**Review by alex-solovyev — APPROVED** [review]
Auto-approved by pulse runner @alex-solovyev — author @vladimirdulov confirmed collaborator, pre-merge gates passed.

---
[aidevops.sh](https://aidevops.sh) v3.20.96 automated scan.

**Comment by vladimirdulov** [comment]
CLAIM_RELEASED reason=worker_complete runner=vladimirdulov ts=2026-06-19T20:04:36Z aidevops_version=3.20.96 opencode_version=1.17.8


---
[aidevops.sh](https://aidevops.sh) v3.20.96 plugin for [OpenCode](https://opencode.ai) v1.17.8 with gpt-5.5 spent 5d 1h and 12,372 tokens on this as a headless worker.

**Review by marcusquinn — APPROVED** [review]
Auto-approved by pulse runner @marcusquinn — author @vladimirdulov confirmed collaborator, pre-merge gates passed.

---
[aidevops.sh](https://aidevops.sh) v3.20.96 automated scan.
